Doc 9216 import data

osfameron opened the pull request:
PR for import developer guide

Docs issue: [DOC-9216](https://issues.couchbase.com/browse/DOC-9216)

osfameron committed:
DOC-9216: Import Data guide

Rewritten draft in more "howto" style.

NB: not been able to test the link to cbimport
(as it's in `backup` repo and not quite sure how to build)
so will need to check this in preview/staging.

maria-robobug commented:
Shouldn't this PR be for the `dev-reorg` branch instead of `release/7.0`?

osfameron commented:
> Shouldn't this PR be for the `dev-reorg` branch instead of `release/7.0`?

Yes indeed! Still draft of course, lemme see if I can repoint it, otherwise will recreate...

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+via the `cbimport` command-line tool shipped with Couchbase Server,
+or using the SDK to script the process.
+
+XXX server/current/tools/cbimport.html
```
The Antora path is xref:tools:cbimport.adoc[cbimport].

Is it possible to include the `cbimport` examples in each of the tab sets below, or does it work too differently from the way the SDKs do? (Either is fine)

Refer to #2275 to see how the `cbc` examples are included alongside the SDK examples.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+
+== Parse the import into an array or stream of records
+
+We now need to parse the input data, either by reading the whole document into memory (if that's possible) or streaming each record in turn.
```
This wording reads more like a tutorial. For a how-to guide it should be more like a procedure, with a title identifying the goal of the procedure, and a list of steps, each of which contains a "do" (imperative statement). And definitely second-person.

So something like this — although it's a bit hard to tell if this is the correct suggestion without having seen the code.

```suggestion
To parse the input data, do one of the following:

* Read the whole document into memory, if possible.

* Stream each record in turn.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
++
+--
+
+We can easily stream a *CSV* file with an external library.
```
```suggestion
To stream a *CSV* file, use the XXX external library.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.js[tags=csvStream,indent=0]
+----
+
+*TSV* files can be processed with the same library:
```
```suggestion
Similarly, to process a *TSV* file, use the XXX external library.
```

I'll allow the "similarly", but it's not really necessary.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.js[tags=tsvStream,indent=0]
+----
+
+*JSON* requires a different library, which streams its output in a `{ key: ..., value: ...}` wrapper, so we need to map the stream into the expected format:
```
```suggestion
To stream a *JSON* file, use the YYY library, and map the stream into the expected `{ key: ..., value: ...}` format.
```

Again, without seeing the code, my suggestion may not be exactly right, but hopefully this conveys the right tone and voice.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.js[tags=jsonStream,indent=0]
+----
+
+The *JSON Lines* format requires the same treatment:
```
```suggestion
Similary, to stream the *JSON Lines* format, use the YYY library, and map the stream into the expected `{ key: ..., value: ...}` format.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+Python::
++
+--
+We can easily parse a *CSV* file with an external library.
```
```suggestion
To parse a *CSV* file, use the XXX.py external library.
```

Tighten up the terminology: is it parse, stream, or process? (I don't know the right answer here, but use the correct term for the SDK and the process you are describing, and be consistent across steps and procedures.)

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::python-sdk:howtos:example$import.py[tags=csv-tsv-import;csvImport,indent=0]
+----
+
+*TSV* files can be processed with the same library:
```
```suggestion
Similary, to process a *TSV* file, use the XXX.py external library.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+----
+
+
+*JSON* requires a different library:
```
```suggestion
To process a *JSON* file, use the YYY.py external library.
```

Also tighten up terminology around "file" or "stream".

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::python-sdk:howtos:example$import.py[tags=json-jsonl-import;jsonImport,indent=0]
+----
+
+The *JSON Lines* format requires the same treatment, with a little processing to get the line-by-line behaviour:
```
```suggestion
Similary, to stream the *JSON Lines* format, use the YYY.py library, and process the stream to get the line-by-line behavior.
```

🇺🇸 spelling

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+--
+Comma Separated Values (.csv) are easily exported from many spreadsheet and database applications.
+
+Typically, the first row is a header row containing the names of the columns within the document.
```
Jus to tighten up the wording for a how-to guide: rather than describing the situations they might encounter in a discursive way, _tell them what to do_ (in so many words). So this might be something like

```suggestion
If required, include a header row containing the names of the columns within the document.
```

osfameron replied:
Ah. CSV lets you do either, but the *code examples* use the header row, because I've made decision that's the best approach for us.

Just checking we can be *opinionated* here? And I don't have to pepper with `NOTE:`s pointing out that you could do it differently...

simon-dew replied:
Hmmm, it's an arguable question at this specific point in the document — are you describing the layout of a typical CSV file here, or instructing users how to lay out a CSV file for import?

Generally the example files should illustrate the instructions, rather than the other way around.

It's fine to be opinionated — in fact you have to be, because we have to decide what procedures we're going to cover, and what steps each procedure should include. The instructions should be clear and concise, so we can only cover the most likely use cases, referring to the reference material for everything else.

The point to bear in mind is that in a how-to guide, we don't know exactly what the user is trying to do. They have a problem they're trying to solve: import some data. So (IMHO) we can't say to them, "our file looks like this, we want to do this". We don't know what their file looks like. We have to give them instructions that will cover as much of their problem as is reasonably possible. The examples to illustrate the procedure and confirm that they're on the right lines. Then refer them to the reference pages for edge cases.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.csv[]
+----
+
+{github}
```
Don't think you need this here? It's a complete example rather than an extract?

simon-dew replied:
You haven't defined the `{github}` attribute in this file anyway. See https://raw.githubusercontent.com/couchbase/docs-server/dev-reorg/modules/guides/pages/select.adoc#L6.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.tsv[]
+----
+
+{github}
```
Don't think you need this

simon-dew replied:
You haven't defined the `{github}` attribute in this file

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.json[]
+----
+
+{github}
```
Don't think you need this

simon-dew replied:
You haven't defined the `{github}` attribute in this file

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.jsonl[]
+----
+
+{github}
```
Don't think you need this here

simon-dew replied:
You haven't defined the `{github}` attribute in this file

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+
+XXX cloud get-started cluster-and-date #import page
+
+[{tabs}]
```
I think you do need the `{github}` link for each example in this tab set, as you are including extracts.

simon-dew replied:
You haven't defined the `{github}` attribute in this file

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+----
+include::dotnet-sdk:howtos:example$Import.cs[tag=connect,indent=0]
+----
+--
```
Include a link to the relevant .NET reference page. See the SDK examples [on this page](https://simon-dew.github.io/docs-site/DOC-9196/server/7.0-reorg/guides/create-index.html).

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+----
+include::java-sdk:howtos:example$Import.java[tag=connect,indent=0]
+----
+--
```
Include a link to the relevant Java reference page.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+----
+include::nodejs-sdk:howtos:example$import.js[tag=connect,indent=0]
+----
+--
```
Include a link to the relevant Node.js reference page.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+----
+include::python-sdk:howtos:example$import.py[tag=connect,indent=0]
+----
+--
```
Include a link to the relevant Python reference page.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+
+We now need to parse the input data, either by reading the whole document into memory (if that's possible) or streaming each record in turn.
+
+[{tabs}]
```
In this tab set, include the `{github}` link for each example, and include a link to the relevant SDK reference page for each tab.

osfameron replied:
I'm unsure what this *does*... (I'd copy/pasted it above, it renders as a literal string.)

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+Finally we do an insert (or upsert, to update the document with that key, if it exists.)
+
+
+[{tabs}]
```
In this tab set, include the `{github}` link for each example, and include a link to the relevant SDK reference page for each tab.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
++
+--
+
+Node lets us iterate a stream with a simple `for` loop in the same way as an array:
```
```suggestion
To iterate the stream, use a simple `for` loop in the same way as an array.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::nodejs-sdk:howtos:example$import.js[tag=importStream,indent=0]
+----
+
+The processing and upsert are then done in this routine:
```
Rewrite this as an imperative instruction rather than a description. (Sorry I don't know the details, you'll have to supply them.)

```suggestion
Use the PPPP function to process and the UUUU function to upsert each record.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+Python::
++
+--
+For convenience, let's define functions to select the key, and process the value:
```
```suggestion
Define functions to select the key and process the value.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::python-sdk:howtos:example$import.py[tag=process,indent=0]
+----
+
+Now the upsert function can simply call those:
```
```suggestion
Use the upsert function to call the functions you defined.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+include::python-sdk:howtos:example$import.py[tag=upsertDocument,indent=0]
+----
+
+NOTE: The Python SDK offers a set of batch operations which are marked as _volatile_ as of SDK 3.2.3, which may be more efficient. Here's a brief example for CSV:
```
This is a note, so fine as description rather than instruction.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+
+== Inserting the documents
+
+So we have a document, and we know where to put it, let's go!
+But of course, Couchbase is a "Key-Value" store, and the document is the *value*... but what's the key?
+
+So the first step is to determine the key. This could be as simple as extracting the `id` field from the document, or using an incrementing sequence number.
+
+Then we may want to do additional processing, for example calculating fields, or adding metadata about the importer.
+
+Finally we do an insert (or upsert, to update the document with that key, if it exists.)
```
Again, rewrite to be more of a how-to guide than a tutorial. Should be less hand-holdy, and definitely second person ("you do this!" not "we can do this".)

```suggestion
Having processed each imported document, you can insert it into the keyspace.
Couchbase is a key-value store, and the document is the value, so before you can insert the document, you need to determine the key.

To insert an imported document into the keyspace:

. Specify the key. This could be as simple as extracting the `id` field from the document, or using an incrementing sequence number.

. Do any additional processing, for example calculating fields, or adding metadata about the importer.

. Finally, use an insert operation to the insert the document.
Alternatively, if you want to update an existing document with the same key, use an upsert operation to update the existing document.
```

simon-dew reviewed (changes requested):
Had a quick look at this to make sure it's going in the right direction. I haven't checked out the request and built the document so I have not commented on the technical accuracy. Most changes requested around style and tone to make it more of a how-to guide.

osfameron commented:
Thanks @simon-dew!
cbcimport is next up after the SDK things, I'll try to incorporate if it makes sense, otherwise will precede the SDK ones.
Other notes look good, will finish code samples then get to that polish / approach tweak.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,335 @@
+JSON::
++
+--
+JSON (.json) files are especially well suited to import into Couchbase, as it is the default native datatype. As we want to import a number of objects, and a .json file contains only one single value, we will typically provide a .json with an *array* of the values to import.
```
This is a good example of the point I was trying to make above (L46). We don't know whether the user wants to import a number of objects, although it's a good guess that they do.

If you decide that this section is instructional (how to lay out a file for import), you could reword like this:

```suggestion
JSON (.json) files are especially well suited to import into Couchbase, as it is the default native datatype.
A JSON file can only contain one single value.
If you want to import a number of objects, create a JSON file with an *array* of the values to import.
```

simon-dew replied:
Alternatively, if you decide that this section is descriptive (of the layout of a typical JSON file), you could reword like this:

```suggestion
JSON (.json) files are especially well suited to import into Couchbase, as it is the default native datatype.
A JSON file can contains only one single value.
If you need to import a number of objects,  the JSON file will typically contain an *array* of the values to import.
```

I'll leave you to decide which approach to take for this particular section, consistently of course 😀 , but for the most part I think you should default to the instructional approach for the how-to guides.

osfameron replied:
Though giving an `if` suggests that users may later ask for an `else`... I can say 

> To let you import one or any number of objects as needed, create an array...

Also, it's possible (though unlikely) that they may want to store arrays, strings, etc.
Which is fine (but the "process by adding an import key" example later would then break)

Um, just thinking out loud, I think, will play around and submit again when ready for review!

osfameron committed:
DOC-9216: Import Data - Java Reactive code samples

Introduce and include code for the latest Reactor samples.

Update {github} admonition (location, and also trying with NOTE:
as I think it's hard to see otherwise? But can revert that if
needed as it's not how the other guides had done it.)

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+xref:tools:cbimport.adoc[cbimport] command-line tool shipped with Couchbase Server,
+or using the SDK to script the process.
+
+Data load essentially consists of:
```
When a list of bullets or numbered points is introduced by an incomplete lead-in clause like this, I think the syntax of the points should follow naturally from the lead-in clause. As it is, it reads a bit awkwardly: "Data load essentially consists of prepare data..."

Either update the numbered points to start with a present participle ("Preparing...", "Parsing...", "Connecting...", etc.), or change the lead-in paragraph to a complete sentence:

```suggestion
Data load essentially consists of the following steps:
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+Data load essentially consists of:
+
+. Prepare data in some well known format such as Comma Separated Values or JSON documents. 
+  * Parse this data, and iterate over each document.
```
Not sure why this is a sub-point? Seems equally important

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+. Connect to the appropriate bucket, scope, and collection.
+
+. Decide on the key for this document (could be an ID, a sequence number, or some combination of fields).
+  * Do any additional processing if required.
```
Again, not sure why this is a sub-point and not a top-level list item?

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+--
+JSON Lines (.json) also known as NDJSON is a common format for streaming JSON, with one JSON object per line of text.
+
+[source,jsonl]
```
`jsonl` isn't currently supported by our syntax highlighter, so I would change this to json

```suggestion
[source,json]
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+<.> import a local CSV file
+<.> replace these values with your connection details
+<.> replace the bucket, scope, and collection as required
+<.> you need to choose the ID to import as. This generator will generate IDs such as `airline_1234`
```
Use sentence case and full stop for callout list items. Otherwise it looks weird when a sentence break occurs in the middle of the list item, as in point 4 here.

```suggestion
<.> Import a local CSV file.
<.> Replace these values with your connection details.
<.> Replace the bucket, scope, and collection as required.
<.> Specify an ID for the imported documents. This generator will generate IDs such as `airline_1234`.
```

Also tightened up the wording slightly for point four.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+pass:[.NET]::
++
+--
+For *CSV* and *TSV*, use the xref:https://joshclose.github.io/CsvHelper/[CsvHelper] library: 
```
This has to be a link, not a cross-reference, because it's an external URL. I tend to use `^` for external links so that they open in a new tab.

```suggestion
For *CSV* and *TSV*, use the https://joshclose.github.io/CsvHelper/[CsvHelper^] library: 
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+<.> import a local JSON file
+<.> replace these values with your connection details
+<.> replace the bucket, scope, and collection as required
+<.> you need to choose the ID to import as. This generator will generate IDs such as `airline_1234`
```
Use sentence case and full stop for callout list items. Otherwise it looks weird when a sentence break occurs in the middle of the list item, as in point 4 here.

```suggestion
<.> Import a local JSON file.
<.> Replace these values with your connection details.
<.> Replace the bucket, scope, and collection as required.
<.> Specify an ID for the imported documents. This generator will generate IDs such as `airline_1234`.
```

Also tightened up the wording slightly for point four.

simon-dew replied:
To be clear, I think it's fine to have list items that are _not_ capitalized and do not end with a full stop, if they are all short phrases. But if any list item contains a sentence break, then I think _all_ the list items should be punctuated as complete sentences. The key thing is consistency.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::dotnet-sdk:howtos:example$Import.csx[tags=importTSV,indent=0]
+----
+
+For *JSON* or *JSONL* format, use xref:https://www.newtonsoft.com/json[Newtonsoft]: 
```
This has to be a link, not a cross-reference, because it's an external URL. I tend to use `^` for external links so that they open in a new tab.

```suggestion
For *JSON* or *JSONL* format, use https://www.newtonsoft.com/json[Newtonsoft^]: 
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+Java::
++
+--
+For *CSV* and *TSV*, use the xref:http://opencsv.sourceforge.net/[opencsv] library: 
```
Likewise, this has to be a link, not a cross-reference.

```suggestion
For *CSV* and *TSV*, use the http://opencsv.sourceforge.net/[opencsv^] library: 
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::java-sdk:howtos:example$Import.java[tags=importTSV;!omit,indent=0]
+----
+
+If you are using the *Reactor API* then, as OpenCSV doesn't have a built-in converter, use the Flux::generate method to convert the CSV or TSV file into a stream:
```
Add link to documentation for the `Flux::generate` method

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+using an SDK gives you more flexibility and control.
+However all the same considerations apply, so let us look at those in turn.
+
+=== Parse the import into an array or stream of records
```
The procedures and examples in this section are quite long. For better readability I would split this section into three:

* Parse a CSV or TSV file
* Parse a JSON file
* Parse a JSONL file

With tab sets for each of the four SDKs in each section.

osfameron replied:
OK, CSV/TSV and JSON/JSONL (e.g. 2 sections) is a much simpler edit, and makes more sense to me.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::java-sdk:howtos:example$Import.java[tags=importJSON,indent=0]
+----
+
+If you are using the *Reactor API* then, once you've read the JSON array, use the Flux::fromIterable method to convert it into streams:
```
Add link to documentation for the `Flux::fromIterable` method

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::java-sdk:howtos:example$Import.java[tags=importJSONL;!omit,indent=0]
+----
+
+If you are using the *Reactor API* then open the JSONL file as a stream using the Flux::using method:
```
Add link to documentation for the `flux:using` method

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+Node.js::
++
+--
+For *CSV* and *TSV*, use the xref:https://csv.js.org/parse/[csv-parse] library:
```
This should be a link, not an xref

```suggestion
For *CSV* and *TSV*, use the https://csv.js.org/parse/[csv-parse^] library:
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::nodejs-sdk:howtos:example$import.js[tags=tsvStream,indent=0]
+----
+
+For *JSON* or *JSONL* format, use the xref:https://github.com/uhop/stream-json[stream-json] library.
```
This should be a link, not an xref

```suggestion
For *JSON* or *JSONL* format, use the https://github.com/uhop/stream-json[stream-json] library.
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+Python::
++
+--
+For *CSV* and *TSV*, use the xref:https://docs.python.org/3/library/csv.html[csv] library:
```
Link not xref

```suggestion
For *CSV* and *TSV*, use the https://docs.python.org/3/library/csv.html[csv^] library:
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::python-sdk:howtos:example$import.py[tags=tsvImport,indent=0]
+----
+
+For *JSON* and *JSONL*, use the xref:https://docs.python.org/3/library/json.html[json] library:
```
Link not xref

```suggestion
For *JSON* and *JSONL*, use the https://docs.python.org/3/library/json.html[json^] library:
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+
+First, you need the connection details for the Couchbase server.
+
+Now decide which bucket and xref:learn:data:scopes-and-collections.adoc[scope and collection] you want to import to (and create them if they don't already exist.)
```
Broken xref

```suggestion
Now decide which bucket and xref:learn:data/scopes-and-collections.adoc[scope and collection] you want to import to (and create them if they don't already exist.)
```

Hint: always have `nav.adoc` open in your editor. You can copy and paste xrefs from there to make sure you get the Antora xref co-ordinates right.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::java-sdk:howtos:example$Import.java[tags=connect,indent=0]
+----
+
+If you are using the xref:java:howtos:concurrent-async-apis.adoc#reactive-programming-with-reactor[Reactive API], then use the reactive collection instead:
```
Broken xref:

```suggestion
If you are using the xref:java-sdk:howtos:concurrent-async-apis.adoc#reactive-programming-with-reactor[Reactive API], then use the reactive collection instead:
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+Java::
++
+--
+Hook the data prepared above into an `upsert` routine. For the blocking API, uses the method below:
```
Grammatical fix

```suggestion
Hook the imported data into an `upsert` routine. For the blocking API, use the method below:
```

osfameron replied:
It's not "imported" at this point, but rewording throughout for more consistency.. "prepared data".

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::python-sdk:howtos:example$import.py[tags=upsertDocument,indent=0]
+----
+
+Learn more about xref:java-sdk:howtos:kv-operations.adoc[].
```
Should this be python?

```suggestion
Learn more about xref:python-sdk:howtos:kv-operations.adoc[].
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+
+The *Reactive API* examples above already include a call to `ReactiveCollection::upsert`.
+
+In both cases, you must provide a preprocess routine which returns a key/value tuple object:
```
Style guide says [key-value](https://github.com/couchbase/docs-site/blob/master/home/modules/contribute/pages/word-list.adoc#k)

```suggestion
In both cases, you must provide a preprocess routine which returns a key-value tuple object:
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::java-sdk:howtos:example$Import.java[tags=upsertDocument,indent=0]
+----
+
+The *Reactive API* examples above already include a call to `ReactiveCollection::upsert`.
```
I won't keep adding comments on all of these, but as a general rule, when you tell users to use a particular Couchbase method, please add a link to the reference documentation for that method. (I know, some SDKs have better / more complete reference documentation than others...)

osfameron replied:
Ah, the Flux ones earlier were an external library, but yes, good point!

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+pass:[.NET]::
++
+--
+Hook the imports above into an `upsertDocument` routine:
```
Vague. Maybe something like "Hook the imported data ..."

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+Node.js::
++
+--
+To iterate the stream you created earlier, use a simple `for` loop in the same way as an array.
```
Vague. "To iterate the imported data stream..."

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+include::python-sdk:howtos:example$import.py[tags=process,indent=0]
+----
+
+Now hook the imports prepared earlier into an `upsertDocument` routine which uses these functions.
```
"Now hook the imported data...."

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+====
+
+
+== Next Steps
```
This is a self-contained how-to guide so there are no next steps.

```suggestion
== Related Links
```

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+
+See also:
+
+* XDCR
```
Can you add links to the `kv-operations` page for every SDK? Copy from @maria-robobug 's Creating Data how-to guide.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+
+// tag::clients_section[]
+// include::partial$clients.adoc[]
+// following section copied/tweaked from above
+// differences: cbc vs cbimport, topic, etc.
```
If necessary you can tweak the partial so it applies to your use-case as well as existing use-cases?

osfameron replied:
It needed a little more thought to unpick (copy/tweak absolutely isn't ideal!) but can look at this afterwards.

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+
+// end::clients_section[]
+
+== Preparing the Data
```
There's a lot of inconsistency within this page about headings. So this heading says "Preparing" with a participle, whereas the next heading is "Use cbimport" with a bare imperative verb. Our style guide is silent on this, but think we should at least be consistent within the page. So either change all the headings to "-ing" forms or change this to "Prepare the Data"

simon-dew commented on modules/guides/pages/import.adoc:
```diff
@@ -0,0 +1,613 @@
+--
+====
+
+=== Inserting the documents
```
Either change all the headings to "-ing" forms or change this to "Insert the documents". Also be consistent with casing! Personally I would change all headings to use title case, whereas you have a mixture of title case and sentence case.

simon-dew reviewed (changes requested):
Thanks for the updates @osfameron ! Most of my comments are to do with consistency of style and tone. I've also suggested breaking down the Parse section as the procedures / examples are quite long.

One more comment to add, outside of the changes you have made:

- [x] Add a link to `import.adoc` in the navigation — there's a dummy entry for this page already

osfameron committed:
DOC-9216: Import Data (review steps wip)

Resolved low-hanging fruit

osfameron committed:
DOC-9216: Import Data review comments

 * more consistent title casing
 * additional xref links
 * split parsing into CSV/TSV and JSON/JSONL

osfameron committed:
DOC-9216: import guide nav position

simon-dew commented:
LGTM! 👍
